File: 2022/r/day02.r
# --- Day 2: Rock Paper Scissors ---

input <- read.table("2022/data/input02.txt", sep=" ")

# Part 1
apply(input, 1, \(x) {
  
    opponent = setNames(1:3, c('A','B','C'))[x[1]]
    choice   = setNames(1:3, c('X','Y','Z'))[x[2]]
    
    c(3, 6, 0)[1 + (choice - opponent) %% 3] + choice
  }) |>
  sum()

# Part 2
apply(input, 1, \(x) {
  
    opponent = setNames(1:3, c('A','B','C'))[x[1]]
    strategy = setNames(1:3, c('X','Y','Z'))[x[2]]
    choice = 1 + (opponent + strategy) %% 3
    
    c(0, 3, 6)[strategy] + choice
  }) |>
  sum()

File: 2022/julia/day02.jl
# --- Day 2: Rock Paper Scissors ---

using DelimitedFiles

input = readdlm("2022/data/input02.txt", ' ', Char)

function part1(input)
    score = 0
    for row in eachrow(input)
        opponent = row[1] - 'a'
        you = mod(row[2] - 'a', 3) + 1

        score += [3, 6, 0][1 + mod(you - opponent, 3)] + you
        
    end
    return score
end

println("Answer 1: ", part1(input))

function part2(input)
    score = 0
    for row in eachrow(input)
        opponent = row[1] - 'a'
        you = mod(row[2] - 'a', 3) + 1

        score += [0, 3, 6][you] + (1 + mod(opponent + you, 3))
    end
    return score
end

println("Answer 2: ", part2(input))
